daemons/execd/remoted_schemas.c

```diff
@@ -0,0 +1,229 @@
+                    pcmk_rc_str(rc));
+        }
+
+    } else if (!S_ISDIR(sb.st_mode)) {
+        /* If something exists with the same name that's not a directory, that's
+         * an error.
+         */
+        crm_err("%s already exists but is not a directory", PCMK__REMOTE_SCHEMA_DIR);
+
+    } else {
+        /* It's a directory - clear it out so we can download potentially new
+         * schema files.
+         */
+        rc = nftw(PCMK__REMOTE_SCHEMA_DIR, rm_files, 10, FTW_DEPTH|FTW_MOUNT|FTW_PHYS);
+
+        if (rc != 0) {
+            crm_err("Could not remove %s: %s", PCMK__REMOTE_SCHEMA_DIR, pcmk_rc_str(rc));
+        }
+    }
+}
+
+static void
+write_extra_schema_file(xmlNode *xml, void *user_data)
+{
+    const char *file = NULL;
+    char *path = NULL;
+    int rc;
+
+    file = crm_element_value(xml, PCMK__XA_PATH);
+    if (file == NULL) {
+        crm_warn("No destination path given in schema request");
+        return;
+    }
+
+    path = crm_strdup_printf("%s/%s", PCMK__REMOTE_SCHEMA_DIR, file);
+
+    /* The schema is a CDATA node, which is a child of the <file> node.  Traverse
+     * all children and look for the first CDATA child.  There can't be more than
+     * one because we only have one file attribute on the parent.
+     */
+    for (xmlNode *child = xml->children; child != NULL; child = child->next) {
+        FILE *stream = NULL;
+
+        if (child->type != XML_CDATA_SECTION_NODE) {
+            continue;
+        }
+
+        stream = fopen(path, "w+");
+        if (stream == NULL) {
+            crm_warn("Could not write schema file %s: %s", path, strerror(errno));
+        } else {
+            rc = fprintf(stream, "%s", child->content);
+
+            if (rc < 0) {
+                crm_warn("Could not write schema file %s: %s", path, strerror(errno));
+            }
+
+            fclose(stream);
+        }
+
+        break;
+    }
+
+    free(path);
+}
+
+static void
+get_schema_files(void)
+{
+    int rc = pcmk_rc_ok;
+    cib_t *cib = NULL;
+    xmlNode *reply;
+
+    cib = cib_new();
+    if (cib == NULL) {
+        _exit(ENOTCONN);
+    }
+
+    rc = cib->cmds->signon(cib, crm_system_name, cib_query);
+    if (rc != pcmk_ok) {
+        crm_err("Could not connect to the CIB manager: %s", pcmk_strerror(rc));
+        _exit(pcmk_rc2exitc(rc));
+    }
+
+    /* Clean up any extra schema files we downloaded from a previous cluster
+     * connection.  After the files are gone, we need to wipe them from
+     * known_schemas, but there's no opposite operation for add_schema().
+     *
+     * Instead, unload all the schemas.  This means we'll also forget about all
+     * the installed schemas as well, which means that xml_latest_schema() will
+     * fail.  So we need to load the base schemas right now.
+     */
+    clean_up_extra_schema_files();
+    crm_schema_cleanup();
+    crm_schema_init();
+
+    rc = cib->cmds->fetch_schemas(cib, &reply, xml_latest_schema(), cib_sync_call);
```

Comment:
What happens if the cluster node doesn't support the new command? Ideally we'd check before attempting to fetch, but if older daemons will reply with an error, that would be fine too.

Reply:
We do exchange `LRMD_PROTOCOL_VERSION` in the handshake. We could (mis)use that to determine CIB protocol support.

Background: `LRMD_PROTOCOL_VERSION` was initially intended for "do we allow this remote connection at all". We bumped it by mistake, so we added `LRMD_MIN_PROTOCOL_VERSION` to work around that. That could be useful though -- we could use `LRMD_MIN_PROTOCOL_VERSION` for connection support, and `LRMD_PROTOCOL_VERSION` for specific feature support. Or, maybe an even better approach would be to use the protocol major version for connection compatibility and the minor version for feature support.

Reply:
Added a commit on top to take care of this.  I'm going with the latter approach, where the major version is for connection compatibility and the minor version is for features.